ADO.NET/OAuthSampleApp/OAuthSampleApp.cs

```diff
@@ -0,0 +1,217 @@
+        connectionStringBuilder["OAuthAccessToken"] = accessToken;
+    }
+
+    // OAuth user attempts to connect with the access token
+    // If it fails, we try to get a new access token with the refresh token
+    // If the refresh token is expired, we get new access and refresh tokens
+    public static async Task ConnectToDatabase()
+    {
+        // ADO.NET requires a user in the connection string
+        // It is ignored by the driver, and only the access token is used
+        // The user can be set to any value (except the admin) and it will connect
+        connectionStringBuilder.User = ConfigurationManager.AppSettings["User"];
+        int retryCount = 0;
+
+        while (retryCount <= 1)
+        {
+            try
+            {
+                using (VerticaConnection conn = new VerticaConnection(connectionStringBuilder.ConnectionString))
+                {
+                    Console.WriteLine("Attempting to connect with connection string: " + ResultSetPrinter.printableConnectionString(connectionStringBuilder.ConnectionString));
+                    conn.Open();
+
+                    using (VerticaCommand command = conn.CreateCommand())
+                    {
+                        Console.WriteLine("Connection Successful");
+
+                        // change the query in app.config or make the command text a string literal
+                        command.CommandText = ConfigurationManager.AppSettings["Query"];
+                        using (VerticaDataReader reader = command.ExecuteReader())
+                        {
+                            Console.WriteLine("Executing Query: " + command.CommandText);
+                            ResultSetPrinter printer = new ResultSetPrinter(reader);
+                            printer.printResults();
+                            Console.WriteLine("Query Executed. Exiting.");
+
+                            break;
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.Message);
+                if (retryCount > 0) { break; }
+                try
+                {
+                    Console.WriteLine("Access token invalid or expired. Attempting token refresh.");
+                    await DoTokenRefresh();
+                }
+                catch (HttpRequestException hre)
+                {
+                    Console.WriteLine(hre.Message);
+                    Console.WriteLine("Refresh token invalid or expired. Attempting to get new access and refresh tokens");
+                    await DoPasswordGrant();
+                }
+
+                SetAccessToken();
+                retryCount++;
+            }
+        }
+    }
+
+    // password grant logs into the IDP using credentials in app.config
+    public static async Task DoPasswordGrant()

```

Comment:
nit: prefer more explanatory/obvious method name. Something like "GetTokensWithPasswordGrant". By the name called elsewhere in the code I wouldn't know what DoPasswordGrant meant unless it was accompanied by a comment.

Reply:
Changed it.